flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala

```diff
@@ -473,8 +473,13 @@ class JobManager(
 
       val jobInfo = new JobInfo(client, listeningBehaviour, System.currentTimeMillis(),
         jobGraph.getSessionTimeout)
-
-      submitJob(jobGraph, jobInfo)
+      val jobID = jobGraph.getJobID
+      if (!currentJobs.contains(jobID)) {
```

Comment:
I think you will need some sort of state in order to determine whether the RecoverAllJobs action has completed... or perhaps better would be to also check here the content of `submittedJobGraphs` to see if it contains the jobId. The Akka actor is single-threaded, but the recovery happens over multiple independent Akka messages which could be interleaved with job submission messages.